spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -157,34 +170,38 @@ public DatastoreEntityConverter getDatastoreEntityConverter() {
 
 	@Override
 	public <T> void deleteById(Object id, Class<T> entityClass) {
-		getDatastoreReadWriter().delete(getKeyFromId(id, entityClass));
+		performDelete(new Key[] { getKeyFromId(id, entityClass) }, Collections.singletonList(id), null, entityClass);
 	}
 
 	@Override
 	public <T> void deleteAllById(Iterable<?> ids, Class<T> entityClass) {
-		List<Key> keys = getKeysFromIds(ids, entityClass);
-		getDatastoreReadWriter().delete(keys.toArray(new Key[keys.size()]));
+		performDelete(getKeysFromIds(ids, entityClass).toArray(new Key[0]), ids, null, entityClass);
 	}
 
 	@Override
 	public <T> void delete(T entity) {
-		getDatastoreReadWriter().delete(getKey(entity, false));
+		performDelete(new Key[] { getKey(entity, false) }, null, Collections.singletonList(entity), entity.getClass());
 	}
 
 	@Override
 	public <T> void deleteAll(Iterable<T> entities) {
-		getDatastoreReadWriter()
-				.delete(StreamSupport.stream(entities.spliterator(), false)
-				.map((x) -> getKey(x, false)).toArray(Key[]::new));
+		performDelete(StreamSupport.stream(entities.spliterator(), false)
```

Comment:
Here you should probably generate several events, one for each entityClass present in the collection.

Reply:
I don't think that would be appropriate. All of the deletes happen in one instant/timestamp for Datastore, which I think would be very confusing if that single operation corresponded to multiple events. It might be impossible to connect the separate events later on to realize they're from the same delete operation.

Reply:
We then have a problem with the definition of an event where we provide the entityClass since it may not be always be there. Maybe make the event just have the keys then. We can always add events that are subclasses of that event to provide more fields where appropriate.

Reply:
@meltsufin  Potentially null fields in events does occur in the other modules. 
In cassandra the table name for the delete event may be null https://github.com/spring-projects/spring-data-cassandra/blob/master/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/event/AbstractDeleteEvent.java

in JDBC the base relational event has a nullable mutation field !
https://github.com/spring-projects/spring-data-jdbc/blob/master/spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/event/SimpleRelationalEvent.java

Reply:
@meltsufin you may want to sit down for this one:
https://github.com/spring-projects/spring-data-jdbc/blob/master/spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/event/BeforeDeleteEvent.java

JDBC delete event has an OPTIONAL entity field!

Reply:
Thanks for the examples. Yes, they do have at most one nullable field. I'm not sure of their reasons for it. Sometimes it's a compromise, but I think in our case we have a few more. Maybe it would be more palatable if we just had one nullable field.

Reply:
While there are 3 fields that are sometimes null I must mention they're never all null at the same time. Like the examples usually just 1 field is null.  

Let me check with the Spring Data guys about this.

Reply:
@mp911de 

We have here just one pair of Before/After DeleteEvent here based on looking at the JDBC and Cassandra modules. However based on the specific delete request sometimes 1 of the fields in the event is NULL. (sometimes the entity, sometimes ID, sometimes entity type). 
How should this be handled?  While there are 3 nullable fields here I must note that they are NEVER all optional. Usually only 1 of them is null.

Reply:
There's not really standard across events. This is due to the various implementations of how stores work (enriching native driver methods with Spring functionality) and how they treat deletes. In some cases, we just propagate the id, sometimes a query and sometimes the entity itself.

MongoDB is quite convenient as a delete by id `Query`/delete by query `Query` is always a document and so we can pass on always the query document. In Cassandra, we pass the `Statement` object as `BeforeDeleteEvent`.
In JDBC, an entity is comprised potentially of multiple tables so a single delete statement does not work there.

Does this help?